src/AggregateFunctions/AggregateFunctionKolmogorovSmirnovTest.h

```diff
@@ -0,0 +1,323 @@
+                         [](const Float64 & q, const Float64 & r, const Float64 & s) { return fabs(r - s) >= q; }
+                       : [](const Float64 & q, const Float64 & r, const Float64 & s) { return r - s >= q; };
+
+            u[0] = 0;
+            for (UInt64 j = 1; j <= n1; j++)
+                if (check(k_q, 0., j / f_n1))
+                    u[j] = 1.;
+                else
+                    u[j] = u[j - 1];
+
+            for (UInt64 i = 1; i <= n2; i++)
+            {
+                if (check(k_q, i / f_n2, 0.))
+                    u[0] = 1.;
+                for (UInt64 j = 1; j <= n1; j++)
+                    if (check(k_q, i / f_n2, j / f_n1))
+                        u[j] = 1.;
+                    else
+                    {
+                        Float64 v = i / static_cast<Float64>(i + j);
+                        Float64 w = j / static_cast<Float64>(i + j);
+                        u[j] = v * u[j] + w * u[j - 1];
+                    }
+            }
+            p_value = u[n1];
+        }
+        else if (method == "asymp")
+        {
+            Float64 n = std::min(n1, n2);
+            Float64 m = std::max(n1, n2);
+            Float64 p = sqrt((n * m) / (n + m)) * d;
+
+            if (alternative == Alternative::TwoSided)
+            {
+                /* reference:
+                 * J.DURBIN
+                 * Distribution theory for tests based on the sample distribution function
+                 */
+                Float64 new_val, old_val, s, w, z;
+                UInt64 k_max = static_cast<UInt64>(sqrt(2 - log(tol)));
+
+                if (p < 1)
+                {
+                    z = - (M_PI_2 * M_PI_4) / (p * p);
+                    w = log(p);
+                    s = 0;
+                    for (UInt64 k = 1; k < k_max; k += 2)
+                        s += exp(k * k * z - w);
+                    p = s / 0.398942280401432677939946059934;
+                }
+                else
+                {
+                    z = -2 * p * p;
+                    s = -1;
+                    UInt64 k = 1;
+                    old_val = 0;
+                    new_val = 1;
+                    while (fabs(old_val - new_val) > tol)
+                    {
+                        old_val = new_val;
+                        new_val += 2 * s * exp(z * k * k);
+                        s *= -1;
+                        k++;
+                    }
+                    p = new_val;
+                }
+                p_value = 1 - p;
+            }
+            else
+            {
+                /* reference:
+                 * J. L. HODGES, Jr
+                 * The significance probability of the Smirnov two-sample test
+                 */
+
+                // Use Hodges' suggested approximation Eqn 5.3
+                // Requires m to be the larger of (n1, n2)
+                Float64 expt = -2 * p * p - 2 * p * (m + 2 * n) / sqrt(m * n * (m + n)) / 3.0;
```

Comment:
The most understandable formula..

Reply:
Formula  5.3 on page 485 of The significance probability of the Smirnov two-sample test :
$$P = exp(-2z^{2}-\frac{2z}{3} \frac{m+2n}{\sqrt{mn(m+n)}}+0\frac{1}{n})$$